[Clang] [C++26] Expansion Statements (Part 3: Enumerating Expansion Statements)

Sirraide opened the pull request:
This implements Sema and template instantiation for enumerating expansion statements, as well as expansion of expansion statements in general.

Sirraide committed:
[Clang] [C++26] Expansion Statements (Part 3)

Sirraide committed:
[Clang] [C++26] Expansion Statements (Part 3)

erichkeane reviewed (commented):
No comments here, everything looks reasonable, but I've burned out on review of this, so quitting this patch set.  I'll come back when we get the other 2 settled/ready (another time...).

tbaederr commented on clang/lib/Sema/SemaExpand.cpp:
```diff
@@ -65,13 +104,144 @@ StmtResult Sema::ActOnCXXExpansionStmtPattern(
+
+  // The index is a DRE to a template parameter; we should never
+  // fail to evaluate it.
+  Expr::EvalResult ER;
```
I think `EvaluateAsKnownConstInt` would make sense here?

erichkeane reviewed (commented):
1 small question, else i think this LGTM.

erichkeane reviewed (approved):
1 small question, else i think this LGTM.

Sirraide commented on clang/lib/Sema/SemaExpand.cpp:
```diff
@@ -65,13 +104,138 @@ StmtResult Sema::ActOnCXXExpansionStmtPattern(
+    return StmtError();
+
+  // Collect shared statements.
+  SmallVector<Stmt *, 1> Shared;
```
Updated this to 3, which is the maximum number of shared statements you can have

zwuis commented on clang/lib/Sema/SemaTemplateInstantiateDecl.cpp:
```diff
@@ -7108,6 +7138,12 @@ NamedDecl *Sema::FindInstantiatedDecl(SourceLocation Loc, NamedDecl *D,
     // anonymous unions in class templates).
   }
 
+  if (CurrentInstantiationScope) {
+    if (auto Found = CurrentInstantiationScope->getInstantiationOfIfExists(D))
+      if (auto *FD = dyn_cast<NamedDecl>(cast<Decl *>(*Found)))
+        return FD;
+  }
```
I applied this part locally and GH176155 is 'fixed'. But I don't know what the underlying problem is, nor do I know why this fixes it.

Sirraide replied:
> I don't know what the underlying problem is, nor do I know why this fixes it.

Neither do I... I tried a number of things here, but this was the only thing that worked

shafik replied:
Is this a drive by fix? Is https://github.com/llvm/llvm-project/issues/176155 what are talking about being fixed?

Sirraide replied:
Sort of; this change is required for expansion statements to work properly independently of #176155, but from what I recall, the test case introduced by the patch that fixes that issue now compiles instead of erroring (I had to update a test that was inadvertently fixed by one of the patches, and I’m pretty sure it’s that one)

cor3ntin commented on clang/lib/Sema/SemaExpand.cpp:
```diff
@@ -23,6 +23,45 @@
 
 using namespace clang;
 
+// Build a 'DeclRefExpr' designating the template parameter '__N'.
```
Don't we have better terminology for that thing?

Sirraide replied:
Hmm, I can probably change it to ‘Build a 'DeclRefExpr' designating the template parameter that is used as the expansion index’. I don’t think there is any actual terminology for this in the standard; it’s more of an implementation detail

cor3ntin replied:
I think this would be an improvement, thanks

cor3ntin commented on clang/test/Parser/cxx2c-expansion-statements.cpp:
```diff
@@ -10,56 +10,56 @@ struct initializer_list {
+  template for (auto y : {abc, -+, }); // expected-error {{use of undeclared identifier 'abc'}} expected-error {{expected expression}}
   template for (3 : "error") // expected-error {{expansion statement declaration must declare a variable}} \
-                                expected-error {{expansion statement must be a range-based for loop}} expected-error {{TODO (expansion statements)}}
+                                expected-error {{expansion statement must be a range-based for loop}}
```
Can you add test with elements of the list being dependent in different ways?
Can i do 

```
template for (int x; static auto y : {pack});
template for (int x; int y : {pack});
template for (int x; int y : {pack...});
template for (int x; T y : {Dependent});
```

etc ?

Sirraide replied:
I’m pretty sure we have tests for all of these in part 7

cor3ntin replied:
no spoiler! (in all seriousness these stacked PRs are very confusing)

Sirraide replied:
Yeah...

cor3ntin commented on clang/lib/Sema/SemaExpand.cpp:
```diff
@@ -64,13 +103,144 @@ StmtResult Sema::ActOnCXXExpansionStmtPattern(
+
+  assert(CurContext->isExpansionStmt());
+  auto *DS = cast<DeclStmt>(ExpansionVarStmt);
+  if (!DS->isSingleDecl()) {
+    Diag(DS->getBeginLoc(), diag::err_type_defined_in_for_range);
+    return StmtError();
```
We probably want to customize that diagnostic

Sirraide replied:
Hmm, well actually that diagnostic says ‘types may not be defined in a for range declaration’, and in the standard, _for-range-declaration_ is a grammar production that is also used for expansion statements; we can change to ‘in the declaration of an expansion variable’ maybe, but I’m not sure if that’s better

cor3ntin commented on clang/lib/Sema/SemaExpand.cpp:
```diff
@@ -64,13 +103,144 @@ StmtResult Sema::ActOnCXXExpansionStmtPattern(
+
+  assert(CurContext->isExpansionStmt());
+  auto *DS = cast<DeclStmt>(ExpansionVarStmt);
+  if (!DS->isSingleDecl()) {
+    Diag(DS->getBeginLoc(), diag::err_type_defined_in_for_range);
+    return StmtError();
```
We probably want to customize that diagnostic

cor3ntin commented on clang/lib/Sema/SemaExpand.cpp:
```diff
@@ -64,13 +103,144 @@ StmtResult Sema::ActOnCXXExpansionStmtPattern(
+  if (!NumInstantiations)
+    return StmtError();
+
+  // Collect shared statements.
```
Shared?

Sirraide replied:
The ‘shared statements’ are the statements that are ‘shared’ across all expansions of an expansion statement (e.g. the range variable in an iterating expansion statement); see the comment on I think `CXXExpansionStmtPattern` for more information (that or it maybe it was on `CXXExpansionStmtInstantiation`)

cor3ntin commented on clang/lib/Sema/TreeTransform.h:
```diff
@@ -9372,19 +9372,101 @@ TreeTransform<Derived>::TransformCXXForRangeStmt(CXXForRangeStmt *S) {
+  // init-statement) are not in a dependent context and must thus be transformed
+  // in the parent context.
+  auto TransformStmtInParentContext = [&] (Stmt *SubStmt) -> StmtResult {
+    Sema::ContextRAII CtxGuard(SemaRef, SemaRef.CurContext->getParent(),
+                               /*NewThis=*/false);
+    return getDerived().TransformStmt(SubStmt);
```
You are not using the new context object here either?

Sirraide replied:
I think I mentioned that in part 2 just now, but that context object was mainly just for use in `Parser::ParseForStatement()` where we have code paths that are shared between expansion statements and regular for loops; I could use it here as well, but I’d have to move it into a header somewhere that’s accessible to both the parser and sema at that point; I’m not sure it’s worth it but I can do that if you prefer

Sirraide replied:
Also the new context object references `Parser::ForRangeInfo`, which is private, so I’d have to refactor that as well...

shafik commented on clang/include/clang/Basic/DiagnosticSemaKinds.td:
```diff
@@ -5954,6 +5954,8 @@ def note_template_nsdmi_here : Note<
 def note_template_type_alias_instantiation_here : Note<
   "in instantiation of template type alias %0 requested here">;
+def note_expansion_stmt_instantiation_here : Note<
+  "in instantiation of expansion statement requested here">;
```
I don't see this note in a test.

Sirraide replied:
There are several tests with this note in the tests added by the part 7 patch

shafik commented on clang/lib/Sema/SemaExpand.cpp:
```diff
@@ -63,13 +103,144 @@ StmtResult Sema::ActOnCXXExpansionStmtPattern(
+
+  VarDecl *ExpansionVar = dyn_cast<VarDecl>(DS->getSingleDecl());
+  if (!ExpansionVar || ExpansionVar->isInvalidDecl() ||
+      ExpansionInitializer->containsErrors())
```
Can't we move the `ExpansionInitializer->containsErrors()` check further up?

shafik commented on clang/lib/Sema/SemaExpand.cpp:
```diff
@@ -63,13 +103,144 @@ StmtResult Sema::ActOnCXXExpansionStmtPattern(
+
+    // Note that lifetime extension only applies to destructuring expansion
+    // statements, so we just ignore 'LifetimeExtendedTemps' entirely for other
+    // types of expansion statements (this is CWG 3043).
```
I think I would like `CWG 3043` mentioned closer to the beginning of the comment.

Although we are not totally consistent it looks like `CWG3043` would be the preferred spelling.

Sirraide replied:
This comment is not entirely correct anymore given that the current wording of CWG3043 inroduces lifetime extension in enumerating expansion statements (which is planned in a follow-up patch; see #207777). I’ll update the comment.

Sirraide committed:
[Clang] [NFC] Expansion Statements (Part 4: for-range refactor) (#169683)

This is an NFC patch that refactors the range-based `for` code such that
it can be reused for iterating expansion statements. It also adds an
`IsConstexpr` bool parameter to some functions, but this is still
always `false` and currently unused (it will be used by the next patch
in the series).
